app/models/concerns/foreman_puppet/extensions/hostgroup.rb

```diff
@@ -35,6 +36,14 @@ def search_by_config_group(_key, operator, value)
           opts = "hostgroups.id IN(#{hostgroup_ids.join(',')})" if hostgroup_ids.present?
           { conditions: opts }
         end
+
+        def search_by_puppetclass(_key, operator, value)
+          conditions = sanitize_sql_for_conditions(["puppetclasses.name #{operator} ?", value_to_sql(operator, value)])
+          hostgroup_ids = ::Hostgroup.unscoped.with_taxonomy_scope.joins(puppet: :puppetclasses).where(conditions).map(&:subtree_ids)
```

Comment:
The [hostgroup](https://github.com/stejskalleos/foreman/blob/ca731bfe12cfb39e64f1cdb0713eccfbe52548f9/app/models/hostgroup.rb#L52) model is by default scoped under taxonomies, you can do just `::Hostgroup.joins(puppet: :puppetclasses).where(conditions).map(&:subtree_ids)`